src/blockdata/witness.rs

```diff
@@ -0,0 +1,395 @@
+
+                if required_len > MAX_VEC_SIZE {
+                    return Err(self::Error::OversizedVectorAllocation {
+                        requested: required_len,
+                        max: MAX_VEC_SIZE,
+                    });
+                }
+
+                resize_if_needed(&mut content, required_len);
+                element_size_varint
+                    .consensus_encode(&mut content[cursor..cursor + element_size_varint_len])?;
+                cursor += element_size_varint_len;
+                d.read_exact(&mut content[cursor..cursor + element_size])?;
+                cursor += element_size;
+            }
+            content.truncate(cursor);
+            Ok(Witness {
+                content,
+                witness_elements,
+                last,
+                second_to_last,
+            })
+        }
+    }
+}
+
+fn resize_if_needed(vec: &mut Vec<u8>, required_len: usize) {
+    if required_len >= vec.len() {
+        let mut new_len = vec.len().max(1);
+        while new_len <= required_len {
+            new_len *= 2;
+        }
+        vec.resize(new_len, 0);
+    }
+}
+
+impl Encodable for Witness {
+    fn consensus_encode<W: Write>(&self, mut writer: W) -> Result<usize, io::Error> {
+        let len = VarInt(self.witness_elements as u64);
+        len.consensus_encode(&mut writer)?;
+        writer.emit_slice(&self.content[..])?;
+        Ok(self.content.len() + len.len())
+    }
+}
+
+impl Witness {
+    /// Creates [`Witness`] object from an array of byte-arrays
+    pub fn from_vec(vec: Vec<Vec<u8>>) -> Self {
+        let witness_elements = vec.len();
+
+        let content_size: usize = vec
+            .iter()
+            .map(|el| el.len() + VarInt(el.len() as u64).len())
+            .sum();
+        let mut content = vec![0u8; content_size];
+        let mut cursor = 0usize;
+        let mut last = 0;
+        let mut second_to_last = 0;
+        for el in vec {
+            second_to_last = last;
+            last = cursor;
+            let el_len_varint = VarInt(el.len() as u64);
+            el_len_varint
+                .consensus_encode(&mut content[cursor..cursor + el_len_varint.len()])
+                .expect("writers on vec don't errors, space granted by content_size");
+            cursor += el_len_varint.len();
+            content[cursor..cursor + el.len()].copy_from_slice(&el);
+            cursor += el.len();
+        }
+
+        Witness {
+            witness_elements,
+            content,
+            last,
+            second_to_last,
+        }
+    }
+
+    /// Convenience method to create an array of byte-arrays from this witness
+    pub fn to_vec(&self) -> Vec<Vec<u8>> {
+        self.iter().map(|s| s.to_vec()).collect()
+    }
+
+    /// Returns `true` if the witness contains no element
+    pub fn is_empty(&self) -> bool {
+        self.witness_elements == 0
+    }
+
+    /// Returns a struct implementing [`Iterator`]
+    pub fn iter(&self) -> Iter {
```

Comment:
I think the name here is confusing: it is not clear do we iterate over witness stack elements - or witness byte sequence. So I will prefer `fn byte_iter(&self)`

Reply:
We iterate over witness stack elements. It wouldn't be meaningful to iterate over bytes. I don't think there is any ambiguity here.

Reply:
From what I see in the code we do iterate over bytes here: `self.content.iter()` where `content: Vec<u8>` will iterate over bytes.

Reply:
`self.content` is used internally to provide the witness stack elements iterator via `iter()`

Reply:
I agree with @dr-orlovsky here, `chars()` and `bytes()` exist on `&str` for this reason.

Reply:
It is clearly meaningful to iterate over either the characters or bytes of a string. It is not clear to me what it means to iterate over the bytes of a witness, or even in what order they would be returned.

Reply:
OK, if it seems completely obvious that it means stack elements fine for me. I'd just like the docs to be more descriptive.

Just note that non-obvious iterators can be a footgun. I once got burned by `&std::path::Path` implementing `IntoIterator` (Can you guess what it iterates over without looking it up?)